src/com/google/enterprise/adaptor/database/UniqueKey.java

```diff
@@ -160,6 +151,67 @@
     return aclSqlCols;
   }
 
+  /**
+   * Attempt to extract ColumnTypes from the supplied Map
+   * of column names to java.sql.Type
+   *
+   * @param sqlTypes Map of column names to Integer java.sql.Types
+   */
+  void addColumnTypes(Map<String, Integer> sqlTypes) {
+    for (Map.Entry<String, Integer> entry : sqlTypes.entrySet()) {
+      if (types.get(entry.getKey()) != null) {
+        continue;
+      }
+      ColumnType type;
+      switch (entry.getValue()) {
+        case Types.BIT:
+        case Types.BOOLEAN:
+        case Types.TINYINT:
+        case Types.SMALLINT:
+        case Types.INTEGER:
+          type = ColumnType.INT;
+          break;
+        case Types.BIGINT:
+          type = ColumnType.LONG;
+          break;
+        case Types.CHAR:
+        case Types.VARCHAR:
+        case Types.LONGVARCHAR:
+        case Types.NCHAR:
+        case Types.NVARCHAR:
+        case Types.LONGNVARCHAR:
+        case Types.DATALINK:
+          type = ColumnType.STRING;
+          break;
+        case Types.DATE:
+          type = ColumnType.DATE;
+          break;
+        case Types.TIME:
+          type = ColumnType.TIME;
+          break;
+        case Types.TIMESTAMP:
+          type = ColumnType.TIMESTAMP;
+          break;
+        default:
+          String errmsg = "Invalid UniqueKey SQLtype '" + entry.getValue() + "'"
+              + " for " + entry.getKey() + ". Please set an explicit key type"
+              + " in db.uniqueKey.";
+          throw new InvalidConfigurationException(errmsg);
+      }
+      types.put(entry.getKey(), type);
+    }
+    ArrayList<String> badColumns = new ArrayList<>();
+    for (Map.Entry<String, ColumnType> entry : types.entrySet()) {
+      if (entry.getValue() == null) {
+        badColumns.add(entry.getKey());
+      }
+    }
+    if (!badColumns.isEmpty()) {
+      throw new InvalidConfigurationException("Unknown column type for the"
+          + " following columns: " + badColumns);
```

Comment:
At this point, I think you get here when DB connectivity is down, so I was hesitant to use the second sentence. But the more I think about it. No connectivity would result in a SQLException, so you wouldn't get here.